File: internal/node/i2pd.go
package node

import (
	"archive/zip"
	"bytes"
	"context"
	"crypto/sha256"
	_ "embed"
	"encoding/base64"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
	"text/template"
	"time"
)

//go:embed i2pd.conf.tmpl
var i2pdConfTmpl string

type I2pdNode struct {
	t         *testing.T
	Index     int
	DataDir   string
	ConfPath  string
	NS        string
	IP        string
	Port      int
	Floodfill bool
	SAMPort   int // 0 = SAM отключён
	cmd       *exec.Cmd
}

func NewI2pdNode(t *testing.T, index int, ns, ip string, port int, floodfill bool) *I2pdNode {
	dataDir := t.TempDir()
	confPath := filepath.Join(dataDir, "i2pd.conf")

	n := &I2pdNode{
		t:         t,
		Index:     index,
		DataDir:   dataDir,
		ConfPath:  confPath,
		NS:        ns,
		IP:        ip,
		Port:      port,
		Floodfill: floodfill,
	}

	if err := os.MkdirAll(filepath.Join(dataDir, "netDb"), 0755); err != nil {
		t.Fatalf("failed to create netDb dir: %v", err)
	}

	t.Cleanup(func() {
		n.Stop()
	})

	return n
}

type i2pdConfigData struct {
	Host       string
	ZipFile    string
	Port       int
	SAMEnabled bool
	SAMPort    int
}

func (n *I2pdNode) WriteConfig(zipFile string) error {
	tmpl, err := template.New("i2pd").Parse(i2pdConfTmpl)
	if err != nil {
		return err
	}

	data := i2pdConfigData{
		Host:       n.IP,
		ZipFile:    zipFile,
		Port:       n.Port,
		SAMEnabled: n.SAMPort > 0,
		SAMPort:    n.SAMPort,
	}

	var buf bytes.Buffer
	err = tmpl.Execute(&buf, data)
	if err != nil {
		return err
	}

	return os.WriteFile(n.ConfPath, buf.Bytes(), 0644)
}

func (n *I2pdNode) Start() error {
	args := []string{"netns", "exec", n.NS, "i2pd",
		"--datadir=" + n.DataDir,
		"--conf=" + n.ConfPath,
		"--address4=" + n.IP,
		"--loglevel=info",
		"--log=file",
		"--logfile=" + filepath.Join(n.DataDir, "i2pd.log"),
		"--daemon",
	}
	if n.Floodfill {
		args = append(args, "--floodfill")
	}

	cmd := exec.Command("ip", args...)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("failed to start i2pd: %v, output: %s", err, string(out))
	}
	return nil
}

func (n *I2pdNode) RouterInfoPath() string {
	return filepath.Join(n.DataDir, "router.info")
}

func (n *I2pdNode) WaitRouterInfoFloodfill(ctx context.Context) error {
	if !n.Floodfill {
		return nil
	}
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
			data, err := os.ReadFile(n.RouterInfoPath())
			if err != nil {
				continue
			}
			if bytes.Contains(data, []byte("Xf")) {
				return nil
			}
		}
	}
}

// routerInfoI2PHash вычисляет SHA256(RouterIdentity) и возвращает
// строку в I2P base64 (+→-, /→~, без padding).
func routerInfoI2PHash(riPath string) (string, []byte, error) {
	data, err := os.ReadFile(riPath)
	if err != nil {
		return "", nil, fmt.Errorf("read router.info: %v", err)
	}
	if len(data) < 387 {
		return "", nil, fmt.Errorf("router.info too short")
	}
	certLen := int(data[385])<<8 | int(data[386])
	identityLen := 387 + certLen
	if len(data) < identityLen {
		return "", nil, fmt.Errorf("router.info too short for identity")
	}
	h := sha256.Sum256(data[:identityLen])
	b64 := base64.RawStdEncoding.EncodeToString(h[:])
	b64 = strings.ReplaceAll(b64, "+", "-")
	b64 = strings.ReplaceAll(b64, "/", "~")
	return b64, data, nil
}

func (n *I2pdNode) CreateReseedZip(dest string) error {
	return CreateMultiReseedZip(dest, []*I2pdNode{n})
}

// CreateMultiReseedZip создаёт ZIP с RouterInfo нескольких нод.
// Используется для cross-populate: i2pd сам раскладывает файлы по netDb.
func CreateMultiReseedZip(dest string, nodes []*I2pdNode) error {
	f, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer f.Close()
	archive := zip.NewWriter(f)
	defer archive.Close()
	for _, n := range nodes {
		b64, data, err := routerInfoI2PHash(n.RouterInfoPath())
		if err != nil {
			return err
		}
		entry, err := archive.Create(fmt.Sprintf("routerInfo-%s.dat", b64))
		if err != nil {
			return err
		}
		if _, err = entry.Write(data); err != nil {
			return err
		}
	}
	return nil
}

// InstallRouterInfo копирует router.info ноды src в netDb ноды dst
// с правильным именем файла (routerInfo-<hash>.dat).
func InstallRouterInfo(src, dst *I2pdNode) error {
	b64, data, err := routerInfoI2PHash(src.RouterInfoPath())
	if err != nil {
		return err
	}
	// i2pd ≥ 2.60 использует схему netDb/r<первый_char_hash>/routerInfo-<hash>.dat
	subdir := filepath.Join(dst.DataDir, "netDb", "r"+b64[:1])
	if err := os.MkdirAll(subdir, 0755); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(subdir, "routerInfo-"+b64+".dat"), data, 0644)
}

// Stop убивает все процессы i2pd с этим datadir и ждёт их реальной гибели.
// i2pd --daemon форкает: PID-файл может содержать PID родителя (который уже умер),
// а не демона. Используем pkill -9 (SIGKILL) + pgrep чтобы убедиться что всё мертво.
func (n *I2pdNode) Stop() {
	exec.Command("pkill", "-9", "-f", "datadir="+n.DataDir).Run()
	// Ждём пока pgrep перестанет находить процессы
	for range 50 {
		time.Sleep(200 * time.Millisecond)
		out, _ := exec.Command("pgrep", "-f", "datadir="+n.DataDir).Output()
		if len(strings.TrimSpace(string(out))) == 0 {
			break
		}
	}
	// Убираем стale PID-файл чтобы следующий запуск не путался
	os.Remove(filepath.Join(n.DataDir, "i2pd.pid"))
}

func (n *I2pdNode) WaitBootstrapped(ctx context.Context) error {
	logFile := filepath.Join(n.DataDir, "i2pd.log")
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
			data, err := os.ReadFile(logFile)
			if err != nil {
				continue
			}
			if strings.Contains(string(data), "NetDbReq: Exploring new") || strings.Contains(string(data), "Tunnel: all tunnels built") {
				return nil
			}
		}
	}
}

File: testbed/src/node/i2pd.rs
use base64::{engine::general_purpose::STANDARD_NO_PAD, Engine as _};
use sha2::{Digest, Sha256};
use std::{
    error::Error,
    fs,
    io::{Read, Write},
    path::{Path, PathBuf},
    process::Command,
    thread,
    time::{Duration, Instant},
};
use tempfile::TempDir;
use zip::write::SimpleFileOptions;

const I2PD_TEMPLATE: &str = include_str!("../../templates/i2pd.conf.tmpl");

pub struct I2pdNode {
    pub index: usize,
    pub data_dir: TempDir,
    pub conf_path: PathBuf,
    pub ns: String,
    pub ip: String,
    pub port: u16,
    pub floodfill: bool,
    pub sam_port: Option<u16>,
}

impl I2pdNode {
    pub fn new(
        index: usize,
        ns: impl Into<String>,
        ip: impl Into<String>,
        port: u16,
        floodfill: bool,
    ) -> Self {
        let data_dir = tempfile::tempdir().expect("create i2pd tempdir");
        fs::create_dir_all(data_dir.path().join("netDb")).expect("create netDb");
        let conf_path = data_dir.path().join("i2pd.conf");
        Self {
            index,
            data_dir,
            conf_path,
            ns: ns.into(),
            ip: ip.into(),
            port,
            floodfill,
            sam_port: None,
        }
    }

    pub fn write_config(&self, zip_file: Option<&Path>) -> Result<(), Box<dyn Error>> {
        let sam_enabled = self.sam_port.is_some();
        let rendered = minijinja::render!(
            I2PD_TEMPLATE,
            host => self.ip,
            zip_file => zip_file.map(|p| p.display().to_string()).unwrap_or_default(),
            port => self.port,
            sam_enabled => sam_enabled,
            sam_port => self.sam_port.unwrap_or(0),
        );
        fs::write(&self.conf_path, rendered)?;
        Ok(())
    }

    pub fn start(&self) -> Result<(), Box<dyn Error>> {
        let mut args = vec![
            "netns".to_string(),
            "exec".to_string(),
            self.ns.clone(),
            "i2pd".to_string(),
            format!("--datadir={}", self.data_dir.path().display()),
            format!("--conf={}", self.conf_path.display()),
            format!("--address4={}", self.ip),
            "--loglevel=info".to_string(),
            "--log=file".to_string(),
            format!(
                "--logfile={}",
                self.data_dir.path().join("i2pd.log").display()
            ),
            "--daemon".to_string(),
        ];
        if self.floodfill {
            args.push("--floodfill".to_string());
        }
        let out = Command::new("ip").args(&args).output()?;
        if !out.status.success() {
            return Err(format!(
                "failed to start i2pd: {}\nstdout:\n{}\nstderr:\n{}",
                out.status,
                String::from_utf8_lossy(&out.stdout),
                String::from_utf8_lossy(&out.stderr)
            )
            .into());
        }
        Ok(())
    }

    pub fn stop(&self) {
        let needle = format!("datadir={}", self.data_dir.path().display());
        let _ = Command::new("pkill").args(["-9", "-f", &needle]).status();
        for _ in 0..50 {
            thread::sleep(Duration::from_millis(200));
            let out = Command::new("pgrep").args(["-f", &needle]).output();
            if out
                .ok()
                .map(|o| String::from_utf8_lossy(&o.stdout).trim().is_empty())
                .unwrap_or(true)
            {
                break;
            }
        }
        let _ = fs::remove_file(self.data_dir.path().join("i2pd.pid"));
    }

    pub fn router_info_path(&self) -> PathBuf {
        self.data_dir.path().join("router.info")
    }

    pub fn wait_router_info_floodfill(&self, timeout: Duration) -> Result<(), Box<dyn Error>> {
        if !self.floodfill {
            return Ok(());
        }
        let deadline = Instant::now() + timeout;
        while Instant::now() < deadline {
            if let Ok(data) = fs::read(self.router_info_path()) {
                if data.windows(2).any(|w| w == b"Xf") {
                    return Ok(());
                }
            }
            thread::sleep(Duration::from_secs(1));
        }
        Err("timeout waiting for floodfill router.info".into())
    }

    pub fn wait_bootstrapped(&self, timeout: Duration) -> Result<(), Box<dyn Error>> {
        let log_file = self.data_dir.path().join("i2pd.log");
        let deadline = Instant::now() + timeout;
        while Instant::now() < deadline {
            if let Ok(data) = fs::read_to_string(&log_file) {
                if data.contains("NetDbReq: Exploring new")
                    || data.contains("Tunnel: all tunnels built")
                {
                    return Ok(());
                }
            }
            thread::sleep(Duration::from_secs(1));
        }
        Err(format!("timeout waiting for node {} bootstrap", self.index).into())
    }

    pub fn create_reseed_zip(&self, dest: &Path) -> Result<(), Box<dyn Error>> {
        create_multi_reseed_zip(dest, &[self])
    }
}

impl Drop for I2pdNode {
    fn drop(&mut self) {
        self.stop();
    }
}

pub fn create_multi_reseed_zip(dest: &Path, nodes: &[&I2pdNode]) -> Result<(), Box<dyn Error>> {
    let file = fs::File::create(dest)?;
    let mut archive = zip::ZipWriter::new(file);
    let options = SimpleFileOptions::default();
    for node in nodes {
        let (b64, data) = router_info_i2p_hash(&node.router_info_path())?;
        archive.start_file(format!("routerInfo-{b64}.dat"), options)?;
        archive.write_all(&data)?;
    }
    archive.finish()?;
    Ok(())
}

fn router_info_i2p_hash(path: &Path) -> Result<(String, Vec<u8>), Box<dyn Error>> {
    let mut data = Vec::new();
    fs::File::open(path)?.read_to_end(&mut data)?;
    if data.len() < 387 {
        return Err("router.info too short".into());
    }
    let cert_len = ((data[385] as usize) << 8) | data[386] as usize;
    let identity_len = 387 + cert_len;
    if data.len() < identity_len {
        return Err("router.info too short for identity".into());
    }
    let hash = Sha256::digest(&data[..identity_len]);
    let b64 = STANDARD_NO_PAD
        .encode(hash)
        .replace('+', "-")
        .replace('/', "~");
    Ok((b64, data))
}
